src/runtime/hexagon_cache_allocator.cpp

```diff
@@ -0,0 +1,165 @@
+                 free(temp2);
+                 temp2 = prev_node;
+                }
+                prev_node = temp2;
+                temp2 = temp2->next;
+        }
+        //Retry one more time after deallocating unused nodes
+        mem = (uint8_t *)HAP_cache_lock(sizeof(char) * size, NULL);
+        prev = prev_node;
+        if (mem == NULL) {
+            free(temp);
+            halide_print(user_context, "HAP_cache_lock failed.\n");
+            halide_mutex_unlock(&hexagon_cache_mutex);
+            return NULL;
+        }
+    } else if (mem == NULL) {
+        halide_print(user_context, "HAP_cache_lock failed.\n");
+        halide_mutex_unlock(&hexagon_cache_mutex);
+        return NULL;
+    }
+    temp->l2memory = (void *)mem;
+    temp->bytes = size;
+    temp->used = true;
+    temp->next = NULL;
+
+    if (prev != NULL) {
+        prev->next = temp;
+    } else if (hexagon_cache_pool == NULL) {
+        hexagon_cache_pool = temp;
+    }
+    halide_mutex_unlock(&hexagon_cache_mutex);
+    return (void*) temp->l2memory;
+}
+
+static inline void hexagon_cache_pool_put(void *user_context, void *cache_mem) {
+    halide_assert(user_context, cache_mem);
+    halide_mutex_lock(&hexagon_cache_mutex);
+    pcache_pool temp = hexagon_cache_pool;
+    bool found = false; 
+    while (!found && (temp != NULL)) {
+        if (temp->l2memory == cache_mem) {
+            temp->used = false;
+            found = true;
+        }
+        temp = temp->next;
+    }
+    halide_mutex_unlock(&hexagon_cache_mutex);
+}
+
+static inline int hexagon_cache_pool_free(void *user_context) {
+    // TODO: Add Mutex locking for access to hexagon_free_pool ( To be Thread safe )
+    halide_mutex_lock(&hexagon_cache_mutex);
+    pcache_pool temp = hexagon_cache_pool;
+    pcache_pool prev = hexagon_cache_pool;
+    int err = QURT_EOK;
+    while (temp != NULL) {
+        if (temp->l2memory != NULL) {
+            err = HAP_cache_unlock(temp->l2memory);
+            if (err != QURT_EOK) {
+                halide_mutex_unlock(&hexagon_cache_mutex);
+                return err;
+            }
+        }
+        prev = temp->next;
+        free(temp);
+        temp = prev;
+    }
+    hexagon_cache_pool = NULL;
+    halide_mutex_unlock(&hexagon_cache_mutex);
+    return QURT_EOK;
+}
+
+WEAK void *halide_hexagon_allocate_from_l2_pool(void *user_context, size_t size) {
+    //TODO Currently option to retry allocation is disabled, we will have to decide if can be
+    //set by user or pipeline.
+    bool retry = false;
+    halide_print(user_context, "halide_hexagon_allocate_from_l2_pool\n");
+    return hexagon_cache_pool_get(user_context, size, retry);
+}
+
+WEAK void halide_hexagon_free_from_l2_pool(void *user_context, void *ptr) {
+    halide_print(user_context, "halide_hexagon_free_from_l2_pool.\n");
+    hexagon_cache_pool_put(user_context, ptr);
+}
+
+WEAK int halide_hexagon_allocate_l2_pool(void *user_context) {
```

Comment:
At the very least, this will probably need a size parameter.

Reply:
Currently it is a dummy function as we don't use this at all. Once this functionality is clear. Can be modified. L2Pool is allocated when we make the first allocation from Pipeline and not before . The pool grows as we make more allocations. While deallocating we have a wrapper class that returns the allocated memory to pool instead of actually freeing them so that the next frame can use them

Reply:
So currently it does not seem like we will be using this function at all.